src/sss_client/ssh/sss_ssh_knownhosts.c

```diff
@@ -0,0 +1,211 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#define _GNU_SOURCE    /* For basename(3) */
+#include <string.h>
+#include <stdio.h>
+#include <talloc.h>
+#include <popt.h>
+#include <signal.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <netdb.h>
+
+#include "util/util.h"
+#include "util/crypto/sss_crypto.h"
+#include "util/sss_ssh.h"
+#include "sss_client/sss_cli.h"
+#include "sss_client/ssh/sss_ssh_client.h"
+
+static errno_t known_hosts(TALLOC_CTX *mem_ctx, const char *domain, const char *host,
+                           struct sss_ssh_ent **_ent, const char **_keyhost)
+{
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+    struct addrinfo ai_hint;
+    struct addrinfo *ai = NULL;
+    char canonhost[NI_MAXHOST];
+    const char *canonname = NULL;
+    const char *keyhost;
+    struct sss_ssh_ent *ent = NULL;
+
+    if (_ent == NULL || _keyhost == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Invalid NULL received\n");
+        ERROR("Internal error\n");
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate tmp_ctx\n");
+        ERROR("Not enough memory\n");
+        return ENOMEM;
+    }
+
+    memset(&ai_hint, 0, sizeof(struct addrinfo));
+    ai_hint.ai_family = AF_UNSPEC;
+    ai_hint.ai_socktype = SOCK_STREAM;
+    ai_hint.ai_protocol = IPPROTO_TCP;
+    ai_hint.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+
+    DEBUG(SSSDBG_FUNC_DATA, "Looking up canonical name for: %s\n", host);
+    ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+    if (ret != EOK) {
+        ai_hint.ai_flags = AI_CANONNAME | AI_NUMERICSERV;
+        ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getaddrinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = ai->ai_canonname;
+        }
+    } else {
+        ret = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                          canonhost, NI_MAXHOST, NULL, 0, NI_NAMEREQD);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getnameinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = canonhost;
+        }
+    }
+    DEBUG(SSSDBG_FUNC_DATA, "Found canonical name: %s\n", canonname);
+
+    /* look up public keys */
+    ret = sss_ssh_get_ent(tmp_ctx, SSS_SSH_GET_HOST_PUBKEYS,
+                          canonname, domain, host, &ent);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_ssh_get_ent() failed (%d): %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    keyhost = talloc_asprintf(tmp_ctx, "%s,%s", canonname, host);
```

Comment:
I think it can be allocated on `mem_ctx` initially.

Reply:
Yes, it can with the code as it is today, just because this is the last allocation.

However, if other operations are later added after the call to `talloc_asprintf()`, it will require to modify back the existing code to allocate it on `tmp_ctx` and later steal this allocation.

Allocating on `tmp_ctx` is also coherent with the rest of the allocations in this function.

There is, of course, the cost/time of stealing the allocation, but this is a command line tool and, in my opinion, this optimization is negligible.

Reply:
ok